### PR TITLE
Change locale path from "locales" to "locales/xwalk"

### PR DESCRIPTION
**Pull request by heke123**
Both Crosswalk and Chrome they place the en-US.pak into folder
"locales" in build of Linux and windows. Potientially there is
conflicts, so it's better to set Crosswalk's locale-path as
"locales/xwalk".
Add a "xwalk_resource_delegate" class which indicates the right
locale pack file to load for Crosswalk.

### CODE REVIEW

**Comment by heke123** [comment]
@rakuco,  I changed the locales path for GYP build first.  If it's ok, I can continue the GN part. Please help to have a look, Thanks.

**Comment by rakuco** [comment]
See my comment in crosswalk-project/chromium-crosswalk#371: it can be dropped provided you implement a Delegate in this pull request. You also need to improve your `git grep` skills, I can clearly see several other locations that assume the locales are in `locales/`.

**Comment by heke123** [comment]
@rakuco  Thanks for your advise.  PR is updated, totally same as how chromecast does.

File: application/common/application_resource.cc [hunk 1]
```diff
@@ -14,7 +14,7 @@ namespace xwalk {
 namespace application {
 namespace {
 const base::FilePath::StringType WGT_LOCALE_DIRECTORY =
-    FILE_PATH_LITERAL("locales");
+    FILE_PATH_LITERAL("locales/xwalk");
```
Comment (rakuco) [inline application/common/application_resource.cc]: You're supposed to check the code you are changing instead of blindingly replacing everything. This looks like the implementation of the wgt spec, so it has nothing to do with this pull request.

File: runtime/app/xwalk_main_delegate.cc [hunk 2]
```diff
@@ -129,8 +129,11 @@ void XWalkMainDelegate::InitializeResourceBundle() {
 #endif
 
 #if !defined(OS_ANDROID)
+  resource_delegate_.reset(new XWalkResourceDelegate());
   ui::ResourceBundle::InitSharedInstanceWithLocale(
-      "en-US", nullptr, ui::ResourceBundle::DO_NOT_LOAD_COMMON_RESOURCES);
+      "en-US", resource_delegate_.get(),
+      ui::ResourceBundle::DO_NOT_LOAD_COMMON_RESOURCES);
+
```
Comment (rakuco) [inline runtime/app/xwalk_main_delegate.cc]: This empty line is unnecessary.

File: runtime/common/xwalk_resource_delegate.cc [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+bool XWalkResourceDelegate::GetRawDataResource(int resource_id,
+                                              ui::ScaleFactor scale_factor,
+                                              base::StringPiece* value) {
+  return false;
+}
+
+bool XWalkResourceDelegate::GetLocalizedString(int message_id,
+                                              base::string16* value) {
```
Comment (rakuco) [inline runtime/common/xwalk_resource_delegate.cc]: Wrong indentation.

File: runtime/common/xwalk_resource_delegate.cc [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+base::RefCountedStaticMemory* XWalkResourceDelegate::LoadDataResourceBytes(
+    int resource_id,
+    ui::ScaleFactor scale_factor) {
+  return NULL;
+}
+
+bool XWalkResourceDelegate::GetRawDataResource(int resource_id,
+                                              ui::ScaleFactor scale_factor,
```
Comment (rakuco) [inline runtime/common/xwalk_resource_delegate.cc]: Wrong indentation.

File: runtime/common/xwalk_resource_delegate.cc [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                                              base::StringPiece* value) {
```
Comment (rakuco) [inline runtime/common/xwalk_resource_delegate.cc]: Wrong indentation.

File: tools/installer/common/installer.include [hunk 6]
```diff
@@ -142,9 +142,9 @@ stage_install_common() {
   fi
 
   # l10n paks
-  cp -a "${BUILDDIR}/locales" "${STAGEDIR}/${INSTALLDIR}/"
-  find "${STAGEDIR}/${INSTALLDIR}/locales" -type f -exec chmod 644 '{}' \;
-  find "${STAGEDIR}/${INSTALLDIR}/locales" -type d -exec chmod 755 '{}' \;
+  cp -a "${BUILDDIR}/locales/xwalk" "${STAGEDIR}/${INSTALLDIR}/"
```
Comment (rakuco) [inline tools/installer/common/installer.include]: You're likely creating a directory called `${STAGEDIR}/${INSTALLDIR}/xwalk` now, not `${STAGEDIR}/${INSTALLDIR}/locales/xwalk`.

File: runtime/common/xwalk_resource_delegate.h [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+class FilePath;
+}
+
+namespace gfx {
+class Image;
+}
+
+namespace xwalk {
+
+// A singleton resource bundle delegate. Primary purpose is to indicate the
```
Comment (rakuco) [inline runtime/common/xwalk_resource_delegate.h]: The explanation here needs to be much more meaningful. Put yourself in the shoes of someone who will look at this code 2 years from now: it is still unclear why the delegate is needed (isn't there code in `ui` doing that automatically? what's wrong with locales in `locales/`?)

File: runtime/app/xwalk_main_delegate.h [hunk 8]
```diff
@@ -35,12 +36,13 @@ class XWalkMainDelegate : public content::ContentMainDelegate {
   content::ContentBrowserClient* CreateContentBrowserClient() override;
   content::ContentRendererClient* CreateContentRendererClient() override;
 
-  static void InitializeResourceBundle();
```
Comment (rakuco) [inline runtime/app/xwalk_main_delegate.h]: Why?

Comment (heke123) [inline runtime/app/xwalk_main_delegate.h]: The member `"resource_delegate_"` is added and referenced in member-function `InitializeResourceBundle()`.  So it cannot be static unless we define` "resource_delegate_"` as static too.  So I refered the similar usage from chromecast/app/cast_main_delegate.h

File: tools/installer/common/installer.include (hunk 6 above)
Comment (heke123) [inline tools/installer/common/installer.include]: Yes, mistake here. I think we don't need to change this file.  Its original logic works fine with "/locales/xwalk".

**Comment by heke123** [comment]
@rakuco   Bother you to have a look again. Thanks.

File: runtime/common/xwalk_resource_delegate.cc [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+XWalkResourceDelegate::XWalkResourceDelegate() {
+  DCHECK(!g_instance) << "Cannot initialize resource bundle delegate twice.";
+  g_instance = this;
+}
+
+XWalkResourceDelegate::~XWalkResourceDelegate() {
+  DCHECK_EQ(g_instance, this);
+  g_instance = NULL;
```
Comment (wuhengzhi) [inline runtime/common/xwalk_resource_delegate.cc]: ditto.

File: runtime/common/xwalk_resource_delegate.cc [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+gfx::Image XWalkResourceDelegate::GetNativeImageNamed(int resource_id) {
+  return gfx::Image();
+}
+
+base::RefCountedStaticMemory* XWalkResourceDelegate::LoadDataResourceBytes(
+    int resource_id,
+    ui::ScaleFactor scale_factor) {
+  return NULL;
```
Comment (wuhengzhi) [inline runtime/common/xwalk_resource_delegate.cc]: ditto.

File: runtime/app/xwalk_main_delegate.cc [hunk 11]
```diff
@@ -18,6 +18,7 @@
 #include "xwalk/runtime/browser/xwalk_runner.h"
 #include "xwalk/runtime/common/logging_xwalk.h"
 #include "xwalk/runtime/common/paths_mac.h"
+#include "xwalk/runtime/common/xwalk_resource_delegate.h"
```
Comment (wuhengzhi) [inline runtime/app/xwalk_main_delegate.cc]: alpha sort?

File: runtime/common/xwalk_resource_delegate.h [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+namespace xwalk {
+
+// In Chrome,the locales path is set as "locales/" which potentially conflicts
+// with it is in Crosswalk. So purpose of XWalkResourceDelegate is to set the
+// locale pack path of Crosswalk as "locales/xwalk/".
+class XWalkResourceDelegate : public ui::ResourceBundle::Delegate {
+ public:
+  // Returns the singleton of delegate. It doesn't create an instance.
```
Comment (rakuco) [inline runtime/common/xwalk_resource_delegate.h]: I don't see this being called anywhere.

File: runtime/common/xwalk_resource_delegate.h [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+class FilePath;
+}
+
+namespace gfx {
+class Image;
+}
+
+namespace xwalk {
+
+// In Chrome,the locales path is set as "locales/" which potentially conflicts
```
Comment (rakuco) [inline runtime/common/xwalk_resource_delegate.h]: you're missing a space before "the locales".

File: runtime/common/xwalk_resource_delegate.h [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+// with it is in Crosswalk. So purpose of XWalkResourceDelegate is to set the
```
Comment (rakuco) [inline runtime/common/xwalk_resource_delegate.h]: "the purpose"

File: xwalk.gyp [hunk 15]
```diff
@@ -632,8 +634,11 @@
       ],
       'copies': [
```
Comment (rakuco) [inline xwalk.gyp]: Are you sure this is still needed (ie. changing the output directory above in `xwalk_strings` is not enough)?

File: runtime/common/xwalk_resource_delegate.cc [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+
+#include "base/files/file_path.h"
+#include "base/path_service.h"
+#include "ui/gfx/image/image.h"
+
+namespace xwalk {
+
+namespace {
+
+XWalkResourceDelegate* g_instance = NULL;
```
Comment (heke123) [inline runtime/common/xwalk_resource_delegate.cc]: @wuhengzhi   I found in xwalk/runtime/  the usage of "nullptr" and "NULL" is half and half.  Is there  any policy to follow when to use nullptr and NULL?
It is a good point, thanks for pointing out this.

File: runtime/app/xwalk_main_delegate.cc (hunk 11 above)
Comment (heke123) [inline runtime/app/xwalk_main_delegate.cc]: yes, my mistake. should be alpha sort.

File: runtime/common/xwalk_resource_delegate.h (hunk 12 above)
Comment (heke123) [inline runtime/common/xwalk_resource_delegate.h]: Yes you are right, it is not referenced anywhere.  When I "copy" this part code from the src/chromecast, I noticed in cast_resource_delegate of the chromecast, the getinstance() is defined but not referenced either.  
My understanding is:  Since it is designed as a singleton pattern by using the g_instance member,  It is good to provide interface to get this instance. Although it is true we don't reference it so far. That's why I didn't remove it.

File: xwalk.gyp (hunk 15 above)
Comment (heke123) [inline xwalk.gyp]: In above `generate_xwalk_string` action, .pak files are generated into subfolder of "gen" first. Then in target "copies" we decide what is the copy destination.
I think it is ok to folllow the old logic,  no matter what the new destination is.

**Comment by heke123** [comment]
@rakuco  @wuhengzhi   I updated the PR based on comments but didn't modify the (1)getinstance() interface and (2)gen_xwalk_string logic in xwalk.gyp.  
Thanks very much.

File: runtime/common/xwalk_resource_delegate.h (hunk 12 above)
Comment (rakuco) [inline runtime/common/xwalk_resource_delegate.h]: Not only are you introducing a dead method that is never called (and thus the singleton pattern is never used), but the `DCHECK`s in the constructor and destructor means Crosswalk will crash every time in debug mode since `g_instance` will be null.

File: xwalk.gyp (hunk 15 above)
Comment (rakuco) [inline xwalk.gyp]: The pak files are being generated into `<(SHARED_INTERMEDIATE_DIR)/xwalk/locales` only because the value of `grit_out_dir` is set to that; what I'm asking is if there's any issue with setting it to `<(PRODUCT_DIR)/locales/xwalk` instead, so that you don't need to copy anything afterwards.

File: runtime/common/xwalk_resource_delegate.cc (hunk 16 above)
Comment (rakuco) [inline runtime/common/xwalk_resource_delegate.cc]: > Is there any policy to follow when to use nullptr and NULL?

See "Null Pointer Constant" in https://chromium-cpp.appspot.com/

File: runtime/common/xwalk_resource_delegate.h (hunk 12 above)
Comment (heke123) [inline runtime/common/xwalk_resource_delegate.h]: In constructor it is `DCHECK(!g_instance) << "Cannot .... twice"` so it won't crash unless we create two instances. But I agree it is not good that we provide singleton interface but never use it in code.
Do you think it is enough just to remove the `GetInstance()` dead method? Or how about totally remove the singleton part to make it a simple class?

Comment (rakuco) [inline runtime/common/xwalk_resource_delegate.h]: Oops, I misread the logic and it's not going to crash indeed. In any case, I think the entire singleton logic can be removed since it's not used.

File: xwalk.gyp (hunk 15 above)
Comment (heke123) [inline xwalk.gyp]: Understand now. I noticed if we change the `grit_out_dir`, not only the `*.pak` files are generated in new destination, but also the new generated folder `grit` which contains` xwalk_string.h`.
So let's just keep the old logic to make it easy.

File: runtime/common/xwalk_resource_delegate.h (hunk 12 above)
Comment (heke123) [inline runtime/common/xwalk_resource_delegate.h]: No problem, I'll remove the singleton.

**Comment by heke123** [comment]
@rakuco  Singleton removed. Please have a review. thanks.

**Comment by rakuco** [comment]
lgtm, thank you for working on this

**Comment by rakuco** [comment]
Since this broke the Crosswalk Windows build, I'll have to revert the PR for now.

**Comment by rakuco** [comment]
Done in #3848

**Comment by rakuco** [comment]
@heke123: fixing this is fairly easy.

What's happening here is that gyp is smart enough to know that when you reference a directory such as `<(PRODUCT_DIR)/locales` as an input for an action and another action (`xwalk_strings`) is known to produce that same directory as an output then the former should depend on the latter.

This is why the `generate_crosswalk_win_zip` action does not have any explicit dependency on `xwalk_strings` or any of the targets that generate the files it depends on (`icudtl.dat`, `osmesa.dll` etc).

The problem with the original patch in this pull request is that the `xwalk_strings` target still generates the locales in `gen/xwalk/locales` while you're now manually copying each file to `locales/xwalk` instead of copying the entire directory, so gyp cannot know that it should depend on that target.

Basically, if you set `grit_out_dir` here to `locales/xwalk` instead of `xwalk/locales` you can then change your `copies` section to use `<(PRODUCT_DIR)/locales/` as the destination and keep the list of `files` unchanged (which has the additional benefit that you do not need to hardcode all the locales you want to copy).

**Comment by heke123** [comment]
@rakuco  If we set `grit_out_dir` to `locales/xwalk`, the `grit/xwalk_strings.h` will be generated into `locales/xwalk `too. Then we have to find some way to move `grit/xwalk_strings.h` back to `gen/xwalk/grit/` OR we have to change the `include_dirs` of some targets who includes `xwalk_strings.h`

So my solution is to change the `"outputs"` into `'<(PRODUCT_DIR)/locales/xwalk'` to make `generate_crosswalk_win_zip` aware the new locale path. Do you think it's ok? PR updated. please have a look.  Big thanks for your help!
